arctic/store/_ndarray_store.py

```diff
@@ -510,13 +572,21 @@ def _do_write(self, collection, version, symbol, item, previous_version, segment
         # chunk and store the data by (uncompressed) size
         chunk_size = int(_CHUNK_SIZE / sze)
 
+        # Used by the forward pointers implementation
+        shas_to_ids, matched_segment_ids = set(), set()
+
         previous_shas = []
-        if previous_version:
-            previous_shas = set([Binary(x['sha']) for x in
-                                 collection.find({'symbol': symbol},
-                                                 projection={'sha': 1, '_id': 0},
-                                                 )
-                                 ])
+        if ARCTIC_FORWARD_POINTERS is FwPointersCfg.DISABLED:
+            if previous_version:
+                previous_shas = set(Binary(x['sha']) for x in
+                                    collection.find({'symbol': symbol}, projection={'sha': 1, '_id': 0}))
+        else:
+            # For forward pointers we always need to run the following query (not only when previous version exists).
+            # Reason is that if only some of the segments might get written, and we re-execute write() via mongo_retry,
+            # the upserted_id(s) won't be a complete list of all segment ObjectIDs
+            shas_to_ids = {x['sha']: x['_id']
+                           for x in collection.find({'symbol': symbol}, projection={'sha': 1, '_id': 1})}
```

Comment:
This won’t be covered by an index, will it? Also looks quite expensive as you’re iterating over every chunk for a symbol - and mongo will thrash the WT cache as it loads each into memory...

Should we be using shas instead?

Reply:
I will check this tomorrow with an explain() and see what happens because _id is quite special, and we have already an index for sha.

If this proves indeed to be expensive, I will consider the switch to shas.

Reply:
Before sharing the explain() results, some benchmarks vs this:
https://github.com/manahl/arctic/blob/master/arctic/store/_ndarray_store.py#L515-L519

![image](https://user-images.githubusercontent.com/3357748/48629493-a74b1900-e9b1-11e8-9d1e-b7aac18ae6a1.png)

I assume the extra time is only for the transfer of the ObjectIDs and execution is acceptably fast.
The +6ms IMHO in almost a second, can be considered negligible.

Will share the explain() findings shortly

Reply:
Interesting, @jamesblackburn  look how the projection changes the index used between the first and the last queries.

This raises some suspicion that WT cache might indeed get polluted.

![image](https://user-images.githubusercontent.com/3357748/48631117-c9df3100-e9b5-11e8-910f-b58280e1f7de.png)

Reply:
Ah...
![image](https://user-images.githubusercontent.com/3357748/48632632-4e7f7e80-e9b9-11e8-998e-dd8139b7f076.png)

Reply:
Just noticed,
```
eqlib._collection.find({'symbol': symbol}, projection={'sha': 1, '_id': 1}).explain()
```
has a FETCH stage, 
while this doesn't 
```
eqlib._collection.find({'symbol': symbol}, projection={'sha': 1, '_id': 0}).explain()
```

It is quite obvious that @jamesblackburn  intuition is correct, we would be polluting the cache, an d it is a good idea to convert to using SHAs instead, or add an index (symbol, sha, _id).

I'd rather use SHA though for obvious reasons